FEXCore/Source/Interface/Core/OpcodeDispatcher/X87.cpp

```diff
@@ -1011,36 +540,203 @@ void OpDispatchBuilder::X87FRSTOR(OpcodeArgs) {
+  }
+
+  Ref HostFlag_CF = _Bfe(OpSize::i64Bit, 1, FCMP_FLAG_LT, Res);
+  Ref HostFlag_ZF = _Bfe(OpSize::i64Bit, 1, FCMP_FLAG_EQ, Res);
+  Ref HostFlag_Unordered = _Bfe(OpSize::i64Bit, 1, FCMP_FLAG_UNORDERED, Res);
+  HostFlag_CF = _Or(OpSize::i32Bit, HostFlag_CF, HostFlag_Unordered);
+  HostFlag_ZF = _Or(OpSize::i32Bit, HostFlag_ZF, HostFlag_Unordered);
+
+  if (WhichFlags == FCOMIFlags::FLAGS_X87) {
+    SetRFLAG<FEXCore::X86State::X87FLAG_C0_LOC>(HostFlag_CF);
+    SetRFLAG<FEXCore::X86State::X87FLAG_C1_LOC>(_Constant(0));
+    SetRFLAG<FEXCore::X86State::X87FLAG_C2_LOC>(HostFlag_Unordered);
+    SetRFLAG<FEXCore::X86State::X87FLAG_C3_LOC>(HostFlag_ZF);
+  } else {
+    // Invalidate deferred flags early
+    // OF, SF, AF, PF all undefined
+    InvalidateDeferredFlags();
+
+    SetRFLAG<FEXCore::X86State::RFLAG_CF_RAW_LOC>(HostFlag_CF);
+    SetRFLAG<FEXCore::X86State::RFLAG_ZF_RAW_LOC>(HostFlag_ZF);
+
+    // PF is stored inverted, so invert from the host flag.
+    // TODO: This could perhaps be optimized?
+    auto PF = _Xor(OpSize::i32Bit, HostFlag_Unordered, _Constant(1));
+    SetRFLAG<FEXCore::X86State::RFLAG_PF_RAW_LOC>(PF);
+  }
+
+  if (PopTwice) {
+    _PopStackDestroy();
+    _PopStackDestroy();
+  } else if ((Op->TableInfo->Flags & X86Tables::InstFlags::FLAGS_POP) != 0) {
+    _PopStackDestroy();
+  }
+}
+
+void OpDispatchBuilder::FTST(OpcodeArgs) {
+  Ref Res = _F80StackTest(0);
+
+  Ref HostFlag_CF = _Bfe(OpSize::i64Bit, 1, FCMP_FLAG_LT, Res);
+  Ref HostFlag_ZF = _Bfe(OpSize::i64Bit, 1, FCMP_FLAG_EQ, Res);
+  Ref HostFlag_Unordered = _Bfe(OpSize::i64Bit, 1, FCMP_FLAG_UNORDERED, Res);
+  HostFlag_CF = _Or(OpSize::i32Bit, HostFlag_CF, HostFlag_Unordered);
+  HostFlag_ZF = _Or(OpSize::i32Bit, HostFlag_ZF, HostFlag_Unordered);
+
+  SetRFLAG<FEXCore::X86State::X87FLAG_C0_LOC>(HostFlag_CF);
+  SetRFLAG<FEXCore::X86State::X87FLAG_C1_LOC>(_Constant(0));
+  SetRFLAG<FEXCore::X86State::X87FLAG_C2_LOC>(HostFlag_Unordered);
+  SetRFLAG<FEXCore::X86State::X87FLAG_C3_LOC>(HostFlag_ZF);
+}
+
+void OpDispatchBuilder::X87OpHelper(OpcodeArgs, FEXCore::IR::IROps IROp, bool ZeroC2) {
+  DeriveOp(Result, IROp, _F80SCALEStack());
+  if (ZeroC2) {
+    SetRFLAG<FEXCore::X86State::X87FLAG_C2_LOC>(_Constant(0));
+  }
+}
+
+void OpDispatchBuilder::X87ModifySTP(OpcodeArgs, bool Inc) {
+  if (Inc) {
+    _IncStackTop();
+  } else {
+    _DecStackTop();
+  }
+}
+
+// Operations dealing with loading and storing environment pieces
+
+// Reconstruct as a constant the Status Word of the FPU.
+// We only track stack top and each of the code conditions (C flags)
+// Top is 3 bits at bit 11.
+// C0 is 1 bit at bit 8.
+// C1 is 1 bit at bit 9.
+// C2 is 1 bit at bit 10.
+// C3 is 1 bit at bit 14.
+// Optionally we can pass a pre calculated value for Top, otherwise we calculate it
+// during the function runtime.
+Ref OpDispatchBuilder::ReconstructFSW_Helper(Ref T) {
+  // We must construct the FSW from our various bits
+  Ref FSW = _Constant(0);
```

Comment:
I optimized this yesterday, please backport the change :)

Reply:
![8xc51s](https://github.com/user-attachments/assets/0ae952e6-7ecf-4a5a-9291-a7779697179a)

Reply:
Sorry about this - fixed.